### PR TITLE
Diagnostic Scope improvements

### PR DESCRIPTION
**Pull request by sourabh1007**
Making following improvements:

- Use of ConcurrentDict in DiagScope - Can be improved by passing the ActivitySource from DiagScopeFactory and giving ownership of ActivitySource to it.

- Change Scope state on the basis of activity state, to support parent-based sampling

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 1]
```diff
@@ -364,6 +338,13 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                     links: GetActivitySourceLinkCollection(),
                     traceparent: _traceparent,
                     tracestate: _tracestate);
+
+                if (activity == null || ((bool?)ActivityExtensions.GetCustomProperty(activity, "IsAllDataRequested") ?? false))
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:342]: `IsAllDataRequested` is a flag on the activity.

`GetCustomProperty` method calls into [Activity.GetCustomProperty](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L1046), what you need is to call [Activity.IsAllDataRequested](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs#L837).

You'd have to add a new method that gets `IsAllDataRequested` - you can use `ActivitySourceHasListeners` as an example on how to do it with reflection+caching for property

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 2]
```diff
@@ -115,6 +89,11 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
         public void Start()
         {
             Activity? started = _activityAdapter?.Start();
+
+            if (started != null)
+            {
+                this.IsEnabled = IsEnabled && started.GetIsAllDataRequestedFlag();
```
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: Should we just set _currentActivity to null if IsAllDataRequested is false? 
/cc @lmolkova

File: sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj [hunk 3]
```diff
@@ -14,6 +14,7 @@
     <PackageReference Include="NUnit" />
     <PackageReference Include="NUnit3TestAdapter" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
+    <PackageReference Include="OpenTelemetry" VersionOverride="1.4.0" />
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/tests/Azure.Core.Tests.csproj:17]: Instead of using VersionOverride, add this to [package.data.props](https://github.com/Azure/azure-sdk-for-net/blob/main/eng/Packages.Data.props#L185) instead in the tests section.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 4]
```diff
@@ -15,22 +14,16 @@
 
 namespace Azure.Core.Pipeline
 {
-    internal readonly struct DiagnosticScope : IDisposable
+    internal struct DiagnosticScope : IDisposable
```
[the remark this answers is not in the document]
Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:17]: because I want to change the `isEnabled `flag on the basis of activity status. And since it was `readonly`, it was not allowing me to do so.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: We don't lose the activities due to sampling when we create it using activity source, we just get activities with `IsAllDataRequested` flag off. Hence, I did the same thing here. Otherwise, I am fine with setting activities as null (it would make this struct `readonly `again)

[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: But IsEnabled doesn't do anything - it isn't used anywhere.

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 5]
```diff
@@ -468,6 +470,38 @@ public void FailedStopsActivityAndWritesExceptionEventActivitySource()
             CollectionAssert.Contains(activity.Tags, new KeyValuePair<string, string>("Attribute2", "2"));
             CollectionAssert.Contains(activity.Tags, new KeyValuePair<string, string>("az.namespace", "Microsoft.Azure.Core.Cool.Tests"));
         }
+
+        [Test]
+        [NonParallelizable]
+        public void OpenTelemetryCompatibilityWithParentBasedSampling()
```
Comment (lmolkova) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:476]: please add a deterministic test instead.

I.e use 
- `AlwaysOff` sampler and check that no activities were created.

OR 

- use parent-based sampler and manually set `00-valid_trace_id-valid_span_id-00` as parentId and expect it to be sampled out. 
 
?

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:95]: Yes we do! https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/src/Shared/MessagingClientDiagnostics.cs#L173

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 6]
```diff
@@ -288,7 +262,15 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _currentActivity = StartActivitySourceActivity();
                 if (_currentActivity != null)
                 {
-                    _currentActivity.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);
+                    if (!(_currentActivity?.GetIsAllDataRequestedFlag() ?? true))
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:265]: instead of stopping and setting current activity to null (which will break context propagation we care about when activities are sampled out), can we instead either:
- add a private `IsAllDataRequested` flag to DiagScope and set it once activity is started
- or use `_currentActivity?.GetIsAllDataRequestedFlag() == true` every time

to guard any expensive operation that DiagScope does with it?

I can see just a couple of places that's needed:
- [MarkFailed](https://github.com/Azure/azure-sdk-for-net/blob/2800d6ecab44c45489a7868bfb6536fef9dc3620/sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs#L367)
- [AddTag](https://github.com/Azure/azure-sdk-for-net/blob/2800d6ecab44c45489a7868bfb6536fef9dc3620/sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs#L203)

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 7]
```diff
@@ -288,7 +267,7 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _currentActivity = StartActivitySourceActivity();
                 if (_currentActivity != null)
                 {
-                    _currentActivity.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);
+                    _currentActivity?.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:270]:
```suggestion
                    _isAllDataRequested = _currentActivity.GetIsAllDataRequestedFlag();
                    _currentActivity.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);
```

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 8]
```diff
@@ -206,6 +180,8 @@ private class ActivityAdapter : IDisposable
             private string? _traceparent;
             private string? _tracestate;
 
+            private bool _isAllDataRequested;
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:183]:
```suggestion
            private bool _isAllDataRequested = true;
```

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 9]
```diff
@@ -348,14 +327,16 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                     _currentActivity.Start();
                 }
 
+                _isAllDataRequested = _currentActivity?.GetIsAllDataRequestedFlag() ?? true;
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:330]:
```suggestion
```

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 10]
```diff
@@ -374,14 +356,17 @@ public void SetStartTime(DateTime startTime)
...
+                        _diagnosticSource?.Write(_activityName + ".Exception", exception);
+                    }
 
-                if (ActivityExtensions.SupportsActivitySource())
-                {
-                    _currentActivity?.SetErrorStatus(exception?.ToString());
+                    if (ActivityExtensions.SupportsActivitySource())
+                    {
+                        _currentActivity?.SetErrorStatus(exception?.ToString());
+                    }
```
Comment (lmolkova) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:369]:
```suggestion
                if (exception != null)
                {
                    _diagnosticSource?.Write(_activityName + ".Exception", exception);
                }

                if (_isAllDataRequested && ActivityExtensions.SupportsActivitySource())
                {
                    _currentActivity?.SetErrorStatus(exception?.ToString());
                }
```

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 11]
```diff
@@ -468,6 +470,109 @@ public void FailedStopsActivityAndWritesExceptionEventActivitySource()
...
+        [Test]
+        [NonParallelizable]
+        public void OpenTelemetryCompatibilityWithTraceIdRatioBasedSampler()
+        {
+            using var _ = SetAppConfigSwitch();
+
+            // Open Telemetry Listener
+            using TracerProvider OTelTracerProvider = Sdk.CreateTracerProviderBuilder()
+                .AddSource($"Azure.*")
+                .SetSampler(new TraceIdRatioBasedSampler(0.1))
```
Comment (lmolkova) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:483]: is this test validating`TraceIdRatioBasedSampler` behavior? If there is some Azure SDK behavior that's not covered by other tests, can we write deterministic tests for it?

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 12]
```diff
@@ -468,6 +470,109 @@ public void FailedStopsActivityAndWritesExceptionEventActivitySource()
...
+                .SetSampler(new AlwaysOffSampler())
+                .Build();
+
+            DiagnosticScopeFactory clientDiagnostics = new DiagnosticScopeFactory("Azure.Clients", "Microsoft.Azure.Core.Cool.Tests", true, true);
+
+            int activeActivityCounts = 0;
+            for (int i = 0; i < 100; i++)
+            {
+                DiagnosticScope scope = clientDiagnostics.CreateScope("ClientName.ActivityName");
+                scope.Start();
```
Comment (lmolkova) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:521]:
```suggestion
                Assert.IsNull(scope.Start());
```

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 13]
```diff
@@ -468,6 +470,109 @@ public void FailedStopsActivityAndWritesExceptionEventActivitySource()
...
+                if (Activity.Current.IsAllDataRequested)
+                    activeActivityCounts++;
+                scope.Dispose();
```
Comment (lmolkova) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:524]:
```suggestion
                scope.Dispose();
                Assert.IsNull(Activity.Current);
```

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs [hunk 14]
```diff
@@ -468,6 +470,109 @@ public void FailedStopsActivityAndWritesExceptionEventActivitySource()
...
+
+            DiagnosticScopeFactory clientDiagnostics = new DiagnosticScopeFactory("Azure.Clients", "Microsoft.Azure.Core.Cool.Tests", true, true);
+
+            int activeActivityCounts = 0;
+            for (int i = 0; i < 5; i++)
+            {
+                DiagnosticScope scope = clientDiagnostics.CreateScope("ClientName.ActivityName");
+                scope.AddLink($"00-6e76af18746bae4eadc3581338bbe8b{i}-2899ebfdbdce904b-00", "foo=bar");
+
+                scope.Start();
```
Comment (lmolkova) [inline sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:552]:
```suggestion
                Assert.IsNull(scope.Start());
```

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 11 above)
Comment (sourabh1007) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:483]: Added 2 more tests as you suggested in prev comments.

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 12 above)
Comment (sourabh1007) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:521]: Start doesn't return anything...why are we asserting is to null? or I am missing something here?

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 14 above)
Comment (sourabh1007) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:552]: Start doesn't return anything...why are we asserting is to null? or I am missing something here?

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: Just curious - is there a reason to prefer adding the IsAllDataRequested guards to each method vs setting to null? It seems like this will become easy to forget as we add new methods in the future.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 9 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:330]: in the line 270, right after starting activity.The it'd apply to Activity source path and would immediately prevent `_currentActivity.AddTag(OpenTelemetrySchemaAttribute, OpenTelemetrySchemaVersion);` from adding any overhead

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 11 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:483]: is this test still needed? does it test anything related to Azure SDK behavior?

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 6 above)
Comment (lmolkova) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: I don't have any preference, as long as we preserve Activity.Current and don't forget to close started one at the end. Agree that it might be easier to do with `_sampledOutActivity`

Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: assigning to new object and having an extra logic in dispose to handle it, sounds like an overhead. In case my test I can add more assertions to make sure we are not adding Tags in sampled out activities.

File: sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs (hunk 11 above)
Comment (sourabh1007) [reply sdk/core/Azure.Core/tests/ClientDiagnosticsTests.Net50.cs:483]: I thought having one extra test to cover one more kind of sampler, should not be a problem as right now diagnostic scope has zero coverage for sampling. Anyway, I am removing it as you said it is not testing anything related to Azure SDK behaviour

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: More tests are always good, but I was thinking more about new methods that we may add in the future to DiagnosticScope where we will need to remember to add this check.

Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs]: if we make sure that new methods are working with sampling then, we won't miss this check.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 15]
```diff
@@ -267,6 +266,17 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _currentActivity = StartActivitySourceActivity();
                 if (_currentActivity != null)
                 {
+                    _isAllDataRequested = _currentActivity.GetIsAllDataRequestedFlag();
+
+                    if (!_isAllDataRequested)
+                    {
+                        _sampleOutActivity = _currentActivity;
+                        _currentActivity.Stop();
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:274]: We shouldn't stop the current activity.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 16]
```diff
@@ -396,6 +401,7 @@ public void Dispose()
                 {
                     _currentActivity.Stop();
                 }
+                _sampleOutActivity?.Stop();
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:404]: This whole method needs to be updated to work off of either activity, e.g.:
```c#
var activity = _currentActivity ?? _sampledOutActivity;
if (activity == null)
{
  return;
}
...
```

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs (hunk 15 above)
Comment (JoshLove-msft) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:274]: Are tests passing currently? Ideally we would have a test that would fail with this change.

Comment (sourabh1007) [reply sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:274]: yes..client diagnostics tests are passing with current change

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 17]
```diff
@@ -288,6 +266,16 @@ public void AddLink(string traceparent, string? tracestate, IDictionary<string,
                 _currentActivity = StartActivitySourceActivity();
                 if (_currentActivity != null)
                 {
+                    _isAllDataRequested = _currentActivity.GetIsAllDataRequestedFlag();
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:269]: This class variable is not needed. We can just use the return of GetIsAllDataRequestedFlag directly.

File: sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs [hunk 18]
```diff
@@ -561,6 +552,25 @@ public static bool IsW3CFormat(this Activity activity)
             return GetTraceStateStringMethod(activity);
         }
 
+        public static bool GetIsAllDataRequestedFlag(this Activity activity)
```
Comment (JoshLove-msft) [inline sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs:555]:
```suggestion
        public static bool GetIsAllDataRequested(this Activity activity)
```

**Comment by JoshLove-msft** [comment]
Looks great! Thanks @sourabh1007!

**Review by JoshLove-msft — APPROVED** [review]

**Review by lmolkova — APPROVED** [review]
LGTM, thank you!!
